content/blog/2023-12-11-rvv-llvm-gisel/index.md

```diff
@@ -0,0 +1,177 @@
+
+[The pipeline of GlobalISel](https://llvm.org/docs/GlobalISel/Pipeline.html#pipeline) composes of four major passes:
+1. [IRTranslate](https://llvm.org/docs/GlobalISel/IRTranslator.html): Converts LLVM IR into [Generic MachineIR (gMIR)](https://llvm.org/docs/GlobalISel/GMIR.html), which has the same data strucures as [MachineIR](https://llvm.org/docs/MIRLangRef.html) but is less constrained. The constraints will be incrementally enforced along with the pipeline, transforming gMIR to MIR in the end.
+2. [Legalize](https://llvm.org/docs/GlobalISel/Legalizer.html): This pass replaces unsupported operations with supported ones. So there will be no illegal instructions after this step.
+3. [Register Bank Select](https://llvm.org/docs/GlobalISel/RegBankSelect.html): This pass assigns the corresponding [Register Banks](https://llvm.org/docs/GlobalISel/GMIR.html#gmir-regbank) to the [generic virtual registers](https://llvm.org/docs/GlobalISel/GMIR.html#gmir-gvregs) in the gMIR. In essence, Register Bank is a framework for grouping and managing different sets of registers in a target architecture. 
+4. [Instruction Select](https://llvm.org/docs/GlobalISel/InstructionSelect.html): This is the final pass, it selects target MIR instructions based on the gMIR instructions. As mentioned above, the gMIR has been constrained enough that it becomes MIR at this point.
+
+A visualization of the pipeline is shown below: ![](pipeline.png)
+
+Please note that Combiner is not in the scope of this blog.
+
+# What Did I Do?
+Let's start with a minimum code generation example that we want to support in LLVM GlobalISel:
+```
+define <vscale x 1 x i8> @vadd_vv_nxv1i8(<vscale x 1 x i8> %va, <vscale x 1 i8> %b) {
+  %vc = add <vscale 1 x i8> %va, %vb
+  ret <vscale x 1 x i8> %vc
+}
+```
+Here, we define a vector-vector add function that takes in two vector arguments, and returns the sum. For the background knowledge of LLVM [vector types](https://llvm.org/docs/LangRef.html#vector-type), it has both fix-length vectors and scalable vectors. We will be focusing on scalable vectors not only because it's more complex and flexible, it also fits into our abovementioned discussion regarding flexible-widths RISC-V vector registers. The form of scalable vectors in LLVM is as follows:
+```
+<vscale x VF x primitive data type>
+```
+where `VF` is the vectorization/grouping factor, it specifies how many vector elements are there in this group; a primitive data type; and `vscale` is a constant multiple of the specified number of vector elements, and `vscale` is unknown at compile time. For example, `<vscale x 4 x i32>` represents a scalable vector, which is a multiple of 4 32-bit integer values.
```

Comment:
VF which stands for VectorizationFactor in the LLVM world has type `ElementCount`, which makes up `vscale x N` for scalable vectors. It would probably be more clear to use `<vscale x M x scalar data type>`. `M` is the minimum nuber of elements in the vector.